specs/eip4844/polynomial-commitments.md

```diff
@@ -414,6 +388,51 @@ def verify_kzg_proof_impl(commitment: KZGCommitment,
     ])
 ```
 
+#### `verify_kzg_proof_multi`
+
+```python
+def verify_kzg_proof_multi(commitments: Sequence[KZGCommitment],
+                           zs: Sequence[BLSFieldElement],
+                           ys: Sequence[BLSFieldElement],
+                           proofs: Sequence[KZGProof]) -> bool:
+    """
+    Verify multiple KZG proofs efficiently.
+    """
+
+    assert len(commitments) == len(zs) == len(ys) == len(proofs)
+
+    # Compute a random challenge. Note that it does not have to be computed from a hash,
+    # r just has to be random.
+    degree_poly = int.to_bytes(FIELD_ELEMENTS_PER_BLOB, 8, ENDIANNESS)
+    num_commitments = int.to_bytes(len(commitments), 8, ENDIANNESS)
+    data = RANDOM_CHALLENGE_KZG_MULTI_DOMAIN + degree_poly + num_commitments
+
+    # Append each polynomial which is composed by field elements
+    for commitment, z, y, proof in zip(commitments, zs, ys, proofs):
+        data += commitment \
+            + int.to_bytes(z, BYTES_PER_FIELD_ELEMENT, ENDIANNESS) \
+            + int.to_bytes(y, BYTES_PER_FIELD_ELEMENT, ENDIANNESS) \
+            + proof
+
+    hashed_data = hash(data)
+    r = hash_to_bls_field(hashed_data + b'\x00')
+    r_powers = compute_powers(r, len(commitments))
+
+    # Verify: e(sum r^i proof_i, [s]) ==
+    # e(sum r^i (commitment_i - [y_i]) + sum r^i z_i proof_i, [1])
```

Comment:
We can make this slightly simpler:
`e(sum r^i proof_i, [s]) == e(sum r^i (commitment_i - [y_i] + z_i proof_i), [1])`
to avoid the extra g1_lincomb and simplify the code.

Here is an impl in c-kzg: https://github.com/ethereum/c-kzg-4844/pull/123/files#diff-ed568929403e623a4b0c74780ccb41dffc9b594000d41d09b21a407fa2126576R1235

Reply:
Hmm I think I retract my suggestion. While simpler in code, the size $t$ MSM that is removed by the above equation is actually paid by doing $t$ scalar mults (to compute $z_i \pi_{i}$) which is more expensive.

Reply:
Yes, I thought about this and it doesn't make sense.